### PR TITLE
Add hessian computation for sparse softmax xent.

### LINKED ISSUE #9368: PreventGradients in SoftmaxCrossEntropyWithLogit ops
SparseSoftmaxCrossEntropyWithLogits cannot take second order gradients (It's not a beautiful hack, and I am not sure how much computational speed up it will bring). 
Adding PreventGradient node in the gradient graph seems to contaminate the computation graph structure, e.g. I am doing custom forward-mode automatic differentiation.
Reference: https://github.com/tensorflow/tensorflow/blob/r1.0/tensorflow/python/ops/nn_grad.py#L334

### PR DESCRIPTION
**Pull request by mknbv**
Closes #9368

Implements second derivative for `_SparseSoftmaxCrossEntropyWithLogits` similar to `_SoftmaxCrossEntropyWithLogits`. Reference: https://github.com/tensorflow/tensorflow/blob/master/tensorflow/python/ops/nn_grad.py#L448

* Tests that no redundant computation is performed when only the first derivative is requested for computation
* Moves `IsZero` method outside of scope of `_SoftmaxCrossEntropyWithLogits` so that it could be used by both `_SoftmaxCrossEntropyWithLogits` and `_SparseSoftmaxCrossEntropyWithLogits`.

### CODE REVIEW

**Comment by tensorflowbutler** [comment]
Nagging Reviewer @ebrevdo: You have been added as a reviewer to this pull request. Please add your review or reassign. It has been 14 days with no activity and the `awaiting review` label has been applied.

**Comment by ebrevdo** [comment]
LGTM; approval conditioned on all tests passing.

**Review by ebrevdo — DISMISSED** [review]

**Comment by MichaelKonobeev** [comment]
@ebrevdo I fixed unused imports and eager mode related errors, not sure if the rest is related.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @yifeif: It has been 15 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by vrv** [comment]
@rthadur  any reason ebrevdo's review isn't enough?  I'm not familiar with this code...

**Comment by rthadur** [comment]
> @rthadur any reason ebrevdo's review isn't enough? I'm not familiar with this code...

@vrv as this is in backlog for a long time , i was wondering if other reviewers can take a look if ebrevdo's is not available,NM i will wait for ebrevdo's review

**Comment by vrv** [comment]
I see -- maybe try pinging him and find out if he can find an alternative reviewer who is familiar with the code?

**Comment by rthadur** [comment]
> I see -- maybe try pinging him and find out if he can find an alternative reviewer who is familiar with the code?

sure , thank you

File: tensorflow/python/ops/nn_grad.py [hunk 1]
```diff
@@ -494,22 +502,28 @@ def IsZero(g):
...
-  return _BroadcastMul(grad_0, sparse_softmax_grad_without_gradient), None
+  # Second derivative is just softmax derivative w.r.t. logits.
+  softmax_grad = op.outputs[1]
+  grad = _BroadcastMul(grad_loss, softmax_grad)
+
+  logits = op.inputs[0]
+  if grad_grad is not None and not _IsZero(grad_grad):
+    softmax = nn_ops.softmax(logits)
+
+    grad += ((grad_grad - array_ops.squeeze(
```
Comment (ebrevdo) [inline tensorflow/python/ops/nn_grad.py]: nit: bring the array_ops.squeeze( down to a newline to improve formatting.

**Review by ebrevdo — COMMENTED** [review]
overall lg.

**Comment by tensorflowbutler** [comment]
Nagging Reviewer @ebrevdo, @ebrevdo, @akshaym, @ebrevdo: You have been added as a reviewer to this pull request. Please add your review or reassign. It has been 14 days with no activity and the `awaiting review` label has been applied.

**Review by ebrevdo — DISMISSED** [review]
approval conditioned on passing tests.

**Comment by rthadur** [comment]
this is merged internally , waiting for auto-merge to happen, thank you

**Comment by ebrevdo** [comment]
This had to be rolled back, unfortunately, because in eager mode the gradient computation accidentally inserts hessian gradients as it can't tell that the input op is a Zeros or ZerosLike.  We'll have to fix this first.
